### DISCUSSION TITLE
New User - scripting Window behaviours in Pekwm  (question)

### DISCUSSION THREAD

**Discussion by hilprl** [pekwm/pekwm#94] [Q&A]
Hi, I have recently started using PekWm and is liking it very much . I am wondering what is the best method to script certain Window behaviors in Pekwm. For instance I would like to write a script to maximize all the windows in a workspace  and re-tile them, something like DWMs monocle mod behavior. Is this possible? Sorry if this has been asked here before. 

Thanks

**Comment by pekdon** [comment]
As I don't know what the DWM mod does I can't comment on that, pekwm-0.1.x has a builting tiling layouter that is no longer part of pekwm-0.2.x.

For pekwm-0.2.x I guess using pekwmctrl or (wmctrl) to do some work is probably possible, especially now with percentages being available in geometry strings.

**Reply by hilprl** [reply]
Thanks for the reply. I will have a look at the pekwmctrl and other settings you mentioned.

**Reply by pclouds** [reply]
Yeah percentage would be nice to line things up, although I don't think it could line up at the exact pixel to not overlap or leave gaps. Perhaps a combination of SetGeometry with snapping would work well (I don't know if snapping works only dragging mouse or with commands too).

**Reply by okraits** [reply]
A combination of the actions `SetGeometry` and `MoveToEdge` can provide some kind of tiling-like behaviour.
Additionally, you can automatically maximize certain windows with autoproperties. Check out the documentation: https://github.com/pekdon/pekwm/blob/master/doc/autoproperties.md

**Reply by hilprl** [reply]
Thanks everyone for your replies. I kind of achieved the behavior I wanted by writing a python script which uses the wmctrl  tool.

**Reply by pekdon** [reply]
> Thanks everyone for your replies. I kind of achieved the behavior I wanted by writing a python script which uses the wmctrl  tool. 

That's nice! Would you mind sharing the script if somebody else wants to do the same?

**Reply by hilprl** [reply]
> That's nice! Would you mind sharing the script if somebody else wants to do the same?

Sure. Let me tidy it up a little bit. I will post it here.

**Reply by hilprl** [reply]
This is the script which I use to imitate the monocle mode behavior like in the tiling window managers. It is not perfect, but works. 

The script uses a json config file to store the current state of mode in each workspace. I have 4 workspaces configured.

Ex - `{"0": false, "1": false, "2": false, "3": false}`

```
#!/usr/bin/python

import subprocess
import os
import json

wstate_json= "/path/to/config" #insert the path to the config file

def getWorkspace():
    """
    Returns:
        string: the current workspace id
    """
    cmd = 'wmctrl -d | grep \*'
    out = subprocess.check_output(["bash", "-c", cmd])
    out_str = out.decode("utf-8")
    return out_str[0]

def getWindowList(wspace):
    """
    Args:
        wspace (string) : current workspace
    Returns: 
        the list of window-ids of the windows opened in the current workspace minus the panel bar
    """
    out = subprocess.check_output(['wmctrl', '-l'])
    out_str = out.decode("utf-8")
    out_lst = out_str.split('\n')[:-2]
    winid_lst = []
    for entry in out_lst:
        win_det = entry.split(" ")
        if win_det[2] == wspace:
            winid_lst.append(win_det[0])
    return winid_lst

def getWinstate(wspace):
    """
    Args:
        wspace (string) : current workspace
    Returns:
        bool: the current state of the monole mode as stored in the json file 
    """
    with open(wstate_json) as json_file:
        data = json.load(json_file)
        return data[wspace]

def writeWinstate(wspace, state):
    """
    This function writes the monocle mode state to the json config file
    Args:
        wspace (string) : current workspace
        state (bool) : current state of the monocle mode 
    """
    with open(wstate_json,"r" ) as json_file:
        data = json.load(json_file)
        data[wspace] = state
    with open(wstate_json,"w") as json_file:
        json.dump(data, json_file)

if __name__== '__main__':
    wspace = getWorkspace()
    winList = getWindowList(wspace)
    windowState = getWinstate(wspace)
    if windowState is False:
        for win in winList:
            command = 'wmctrl -i -r {} -b add,maximized_vert,maximized_horz'.format(win)
            os.system(command)
        writeWinstate(wspace,True)
    else:
        for win in winList:
            command = 'wmctrl -i -r {} -b remove,maximized_vert,maximized_horz'.format(win)
            os.system(command)
        writeWinstate(wspace, False)
```
